Source/Particles/ShapeFactors.H

```diff
@@ -11,166 +11,183 @@
  *  Compute shape factor and return index of leftmost cell where
  *  particle writes.
  *  Specialized templates are defined below for orders 0 to 3.
+ *  Shape factor functors may be evaluated with double arguments
+ *  in current deposition to ensure that current deposited by
+ *  particles that move only a small distance is still resolved.
+ *  Without this safeguard, single and double precision versions
+ *  can give disagreeing results in the time evolution for some
+ *  problem setups.
  */
 template <int depos_order>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shape_factor(amrex::Real* const sx, amrex::Real xint)
+struct Compute_shape_factor
 {
-    return 0;
-}
+    template< typename T >
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    int operator()(T* const sx, T xint) const { return 0; };
+};
 
 /**
  *  Compute shape factor and return index of leftmost cell where
  *  particle writes.
  *  Specialization for order 0
  */
 template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shape_factor <0> (amrex::Real* const sx, amrex::Real xmid){
-
-    using namespace amrex;
-
-    const auto j = static_cast<int>(xmid+0.5_rt);
-    sx[0] = 1.0_rt;
-    return j;
-}
+struct Compute_shape_factor< 0 >
+{
+    template< typename T >
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    int operator()(T* const sx, T xmid) const
+    {
+        const auto j = static_cast<int>(xmid + (T)0.5);
+        sx[0] = (T)1.0;
```

Comment:
```suggestion
        sx[0] = T(1.0);
```

This is what differentiates us from `C` ;-)

Reply:
Done :-)